utils/utils.jsx

```diff
@@ -1066,6 +1066,158 @@ export function getCaretPosition(el) {
+    const copy = copyTextAreaToDiv(textArea);
+    const range = document.createRange();
+    range.setStart(copy.firstChild, start);
+    range.setEnd(copy.firstChild, end);
+    const selection = document.getSelection();
+    selection.removeAllRanges();
+    selection.addRange(range);
+    const rect = range.getClientRects();
+    document.body.removeChild(copy);
+    textArea.selectionStart = start;
+    textArea.selectionEnd = end;
+    textArea.focus();
+    return {
+        x: Math.floor(rect[0].left - textArea.scrollLeft),
+        y: Math.floor(rect[0].top - textArea.scrollTop),
+    };
+}
+
+export function getViewportSize(win) {
+    const w = win || window;
+
+    if (w.innerWidth != null) {
+        return {w: w.innerWidth, h: w.innerHeight};
+    }
+
+    const d = w.document;
+    if (d.compatMode === 'CSS1Compat') {
+        return {w: d.documentElement.clientWidth,
+            h: d.documentElement.clientHeight};
+    }
+    return {w: d.body.clientWidth, h: d.body.clientHeight};
+}
+
+export function offsetTopLeft(el) {
+    if (!(el instanceof HTMLElement)) {
+        return {top: 0, left: 0};
+    }
+    const rect = el.getBoundingClientRect();
+    const scrollLeft = window.pageXOffset || document.documentElement.scrollLeft;
+    const scrollTop = window.pageYOffset || document.documentElement.scrollTop;
+    return {top: rect.top + scrollTop, left: rect.left + scrollLeft};
+}
+
+export function getSuggestionBoxAlgn(textArea) {
+    if (!textArea || !(textArea instanceof HTMLElement)) {
+        return {
+            IsOutOfRightSideViewport: false,
+            pixelsToMove: 0,
+        };
+    }
+    const caretXCoordinateInTxtArea = getCaretXYCoordinate(textArea).x;
+    const viewportWidth = getViewportSize().w;
+
+    // value in pixels used in suggestion-list__content class line 72 file _suggestion-list.scss
+    const suggestionBoxWidth = Constants.SUGGESTION_LIST_MODAL_WIDTH;
+
+    // value in pixels for the offsetLeft for the textArea
+    const txtAreaOffsetLft = offsetTopLeft(textArea).left;
+
+    // textArea padding left of 15px defined in the _post.scss line 392 + 1px from the border
+    const txtAreaPaddingLft = Constants.TEXTAREA_PADDING_LEFT + Constants.TEXTAREA_BORDER_WIDTH;
+
+    const pxToSubstract = getPixelsToSubstract();
+
+    // how many pixels to the right should be moved the suggestion box
+    const pxToTheRight = (caretXCoordinateInTxtArea + txtAreaPaddingLft) - (pxToSubstract);
+
+    // the x coordinate in the viewport of the suggestion box border-right
+    const xBoxRightCoordinate = caretXCoordinateInTxtArea + txtAreaOffsetLft + suggestionBoxWidth;
+
+    const willSuggestionBoxOverflowRight = isBoxOutOfRightSideViewport(viewportWidth, xBoxRightCoordinate);
+    return {
+        IsOutOfRightSideViewport: willSuggestionBoxOverflowRight,
+        pixelsToMove: Math.max(0, Math.floor(pxToTheRight)),
+    };
+}
+
+export function getPixelsToSubstract() {
```

Comment:
Perhaps this could be passed into `getSuggestionBoxAlgn` instead of being calculated as part of it? I could see this sort of code being used elsewhere where we might just want to find the position of the caret for other reasons.

Reply:
Also, this might not be what we want when we're displaying other content in the autocomplete. I'll follow up with UX about this outside of the review.

Reply:
ok, that makes perfect sense. I will add a second parameter to getSuggestionBoxAlgn with a default value of 0, and in this scenario, we grab the resulted value from calling getPixelsToSubstract and pass it to getSuggestionBoxAlgn as a second argument.